Start using hierarchical parallelism to limit vertical loops extents

mwarusz opened the pull request:
<!--
Please add a description of what is accomplished in the PR here at the top:
-->
This PR converts the following Omega modules to use hierarchical parallelism for limiting vertical loops extents:
- Auxiliary state
- Tendencies
- EOS

It includes some minor changes to the vertical coordinate module, but #309 has more substantial updates and needs to be merged first.

<!--
Below are a few things we ask you or your reviewers to kindly check.
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] Documentation:
  * [ ] User's Guide has been updated
  * [x] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [ ] Building
  * [ ] CMake build does not produce any new warnings from changes in this PR
* [ ] Testing
  * [x] CTest unit tests for new features have been added per the approved design.
  * [x] Unit tests have passed. Please provide a relevant CDash build entry for verification.
  * [ ] Polaris test suite has passed

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

mwarusz committed:
Add vert coord to aux state and aux vars

mwarusz committed:
Store pointer to HorzMesh in Tendencies

mwarusz committed:
Add vert coord to tendencies and tendency terms

mwarusz committed:
Add test helpers for setting fields that take vertical bounds

mwarusz committed:
Fix AccumTypeHelper

mwarusz committed:
Refactor reductions in test helpers

mwarusz committed:
Add vertical ranges in auxstate test

mwarusz committed:
Use hierarchical parallelism in auxstate

mwarusz committed:
Add option to set boundary in setX helpers

mwarusz committed:
Change vertical limits in tendencies test

mwarusz committed:
Use hierarchical parallelism in tendencies

mwarusz committed:
Add functions to compute chunk start and length

mwarusz committed:
Add KOKKOS_FUNCTION to getVertBound

mwarusz committed:
Fix vert range bugs in some aux vars

mwarusz committed:
Use a kernel to zero tracer tendencies

mwarusz committed:
Remove NChunks from Tendencies

mwarusz committed:
Store HorzMesh and VertCoord in EOS

mwarusz committed:
Use hierarchical parallelism in EOS

mwarusz committed:
Fix EOS Test

mwarusz committed:
Test VelocityDel2Aux in aux state test

mwarusz committed:
Limit vertical range in Del2DivCell

mwarusz committed:
Remove unused variables from aux state

mwarusz committed:
Update timestepping code to loop only over active elements

mwarusz commented:
I have updated this PR to use hierarchical parallelism in the time stepping code to loop only over active layers.

brian-oneill commented:
Looks good as a first step to integrating hierarchical parallelism in Omega. Ran ctests successfully on Chrysalis, pm-cpu, pm-gpu, and Frontier (cpu & gpu). Thanks @mwarusz !

mwarusz committed:
Use a mesh without boundaries in time stepper test

mwarusz committed:
Read VertCoord stream in aux state and tendencies tests

mwarusz committed:
Remove unused variable in aux state test

mwarusz committed:
Update dev docs

mwarusz committed:
Fix formatting

mwarusz commented:
@brian-oneill 

I rebased this branch and updated the docs. I had to change the mesh used in time stepper tests because of issues with forcing constant `NVertLayers` when the input mesh has land boundaries.

mark-petersen commented:
Ran a quick test. Passes CPU and GPU cTests on perlmutter.

mark-petersen commented:
Thanks @mwarusz and @brian-oneill
